File: test_ff5.py
import pandas as pd
import numpy as np

ff5 = pd.read_csv("Data/Factor_Data/ff5.csv")
ff5.columns = ff5.columns.str.strip()
ff5["Month"] = pd.to_datetime(ff5["Month"].astype(str), format="%Y-%m").dt.to_period("M")
for col in ["SMB", "HML", "WML", "RMW", "CMA"]:
    if col in ff5.columns:
        ff5[col] = pd.to_numeric(ff5[col], errors="coerce")
ff5 = ff5.sort_values("Month").reset_index(drop=True)

start = pd.Period("2003-10", freq="M")
end   = pd.Period("2025-12", freq="M")
ff5_plot = ff5[(ff5["Month"] >= start) & (ff5["Month"] <= end)].copy()

FACTORS = ["SMB", "HML", "WML", "RMW", "CMA"]
print(f"\n{'Factor':<8} {'Ann Ret%':>9} {'Ann Vol%':>9} {'Sharpe':>8} {'T-stat':>8} {'N':>6}")
print("-" * 52)
for col in FACTORS:
    if col not in ff5_plot.columns: continue
    s = ff5_plot[col].dropna()
    if len(s) > 1:
        ret   = s.mean() * 12 * 100
        vol   = s.std()  * np.sqrt(12) * 100
        sr    = s.mean() / s.std() * np.sqrt(12)
        tstat = s.mean() / s.std() * np.sqrt(len(s))
        print(f"{col:<8} {ret:>9.2f} {vol:>9.2f} {sr:>8.3f} {tstat:>8.2f} {len(s):>6}")

File: test_ff5.js
const fs = require('fs');
const content = fs.readFileSync('Data/Factor_Data/ff5.csv', 'utf8');
const lines = content.trim().split('\n');
const headers = lines[0].split(',').map(h => h.trim());
const data = lines.slice(1).map(l => {
    const cols = l.split(',');
    const obj = {};
    headers.forEach((h, i) => obj[h] = cols[i] ? cols[i].trim() : null);
    return obj;
});

const start = "2003-10";
const end = "2025-12";
const filtered = data.filter(d => {
    if (!d.Month) return false;
    const m = d.Month.substring(0, 7);
    return m >= start && m <= end;
});

const factors = ["SMB", "HML", "WML", "RMW", "CMA"];
console.log(`\nFactor   Ann Ret%  Ann Vol%   Sharpe   T-stat      N`);
console.log("-".repeat(52));

factors.forEach(f => {
    const vals = filtered.map(d => parseFloat(d[f])).filter(v => !isNaN(v));
    if (vals.length > 1) {
        const mean = vals.reduce((a, b) => a + b, 0) / vals.length;
        const variance = vals.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / (vals.length - 1);
        const std = Math.sqrt(variance);
        
        const annRet = mean * 12 * 100;
        const annVol = std * Math.sqrt(12) * 100;
        const sr = (mean / std) * Math.sqrt(12);
        const tstat = (mean / std) * Math.sqrt(vals.length);
        
        console.log(`${f.padEnd(8)} ${annRet.toFixed(2).padStart(9)} ${annVol.toFixed(2).padStart(9)} ${sr.toFixed(3).padStart(8)} ${tstat.toFixed(2).padStart(8)} ${vals.length.toString().padStart(6)}`);
    }
});
